## PR Review — Automated (pr-reviewer agent)

**Verdict: REQUEST CHANGES** (cannot submit via GitHub review API on self-authored PRs — see memory note `workforce_solo_dev_blockers.md`)

Three blockers. Each is specific and actionable.

---

### BLOCKER 1: CI is red — the shellcheck job this PR introduces is failing

The new Shellcheck CI job fails on `scripts/lib/boot-config.sh` lines 64–68:

```
SC2317 (info): Command appears to be unreachable.
```

The disable directive added at line 62 is `# shellcheck disable=SC2329` (function never called). The failing code is SC2317 (unreachable commands inside the trap-invoked function body). These are different codes. The disable on the function definition line does not suppress SC2317 on the body lines (64–68). Fix: change to `# shellcheck disable=SC2317` applied inside the function body (or as a block around the entire function including the body).

A PR cannot be approved while the gate it introduces is itself red.

---

### BLOCKER 2: flash-sd.sh removable guard is broken for mmcblk devices

Lines 115–116 of `scripts/flash-sd.sh`:

```bash
dev_base="${dev_name%%p[0-9]*}"   # mmcblk0p1 → mmcblk0  ✓
dev_base="${dev_base%%[0-9]}"     # mmcblk0   → mmcblk   ✗ WRONG
```

`%%[0-9]` is "strip longest trailing suffix matching a single digit." For `mmcblk0` it strips the `0`, yielding `mmcblk`. The script then checks `/sys/block/mmcblk/removable`, which does not exist. The sysfs entry for an SD card on a Pi is at `/sys/block/mmcblk0/removable`. Because the path is missing, the code falls through to the `[WARN]` branch and silently allows flashing without the removable check.

Result: for `/dev/mmcblk0` and `/dev/mmcblk0p1` — the exact device names you'll see on a Pi 5 with an SD card — the removable safety guard is always silently bypassed.

Verified locally:
```
dev_name="mmcblk0p1"
dev_base="${dev_name%%p[0-9]*}"  → mmcblk0
dev_base="${dev_base%%[0-9]}"    → mmcblk       ← wrong
sysfs path checked: /sys/block/mmcblk/removable  ← does not exist → WARN bypass
```

The second strip should be removed entirely. The first strip (`%%p[0-9]*`) already correctly handles `mmcblk0p1 → mmcblk0`. For `sdb1` (no `p` prefix), the correct fix is `%[0-9]*` (shortest suffix, single trailing digits). The cleanest solution:

```bash
dev_base="${dev_name%%p[0-9]*}"   # strip mmcblk/nvme partition suffix (mmcblk0p1 → mmcblk0)
dev_base="${dev_base%[0-9]*}"     # strip sata/usb partition suffix (sdb1 → sdb); no-op if no trailing digits
```

Note: `%` not `%%` on the second line to avoid over-stripping `mmcblk0` → `mmcblk`.

This is a destructive-write safety guard on a tool that erases devices. The bug must be fixed.

---

### BLOCKER 3: Scope — 943 additions across 16 files; 9 files outside issue scope

Issue #110 scope is: `build-image.yml`, shellcheck job in `pr-checks.yml`, `flash-sd.sh`, `dev-deploy.sh`, two docs files. This PR also modifies nine pre-existing scripts with shellcheck fixes/disables (`build-image.sh`, `dev-pull-from-pi.sh`, `init-project.sh`, `boot-config.sh`, `install-arlowe-udev-polkit.sh`, `check.sh`, `verify-third-party.sh`, `install-units.sh`, `arlowe-grow-models.sh`). Convention is atomic PRs under 400 net lines (hard cap 600). This PR is 943 additions.

Recommended split:
- PR A (this issue): CI workflow, flash-sd.sh, dev-deploy.sh, docs. ~750 lines of genuinely new files.
- PR B (new `type:chore` issue): shellcheck fixes to pre-existing scripts. Mechanical, reviewable on its own.

If keeping combined: add a note to the PR description explicitly acknowledging the combined scope so the decision is on record.

---

### Suggestions (non-blocking)

- `dev-deploy.sh` line 69: error message says `flash-sd:` but this is `dev-deploy.sh`.
- `dev-deploy.sh` lines 116, 135: `-o StrictHostKeyChecking=no` on both SSH calls. For a LAN dev tool `accept-new` is safer (auto-accepts on first connect, rejects changed keys after).
- `build-image.yml`: `usimd/pi-gen-action@v1` is a floating tag. Pin to a commit SHA for firmware build reproducibility.
- `build-image.yml`: `SOURCE_DATE_EPOCH` is documented in `phase-6-repro-exceptions.md` as a reproducibility lever but is not set in the workflow itself.

---

### What is correct

- `build-image.yml` has no `pull_request` trigger — confirmed.
- macOS root disk detection (`diskutil info / | awk '/Part of Whole:/'`) — correct.
- Linux root disk detection (`findmnt` + `/proc/mounts` fallback + partition suffix strip via sed) — correct.
- System disk prefix guard on both platforms — sound.
- `dev-deploy.sh --dry-run` gates both rsync and systemctl — true dry-run, no mutations.
- Unit restart ordering matches documented dependency order.
- No secrets or sensitive literals in new files.
- Conventional commit format correct.

--- a/scripts/dev-deploy.sh
+++ b/scripts/dev-deploy.sh
@@ -66,7 +66,7 @@ EOF
 while [[ $# -gt 0 ]]; do
     case "$1" in
         --target)
-            [[ $# -ge 2 ]] || { echo "flash-sd: --target requires a value" >&2; exit 1; }
+            [[ $# -ge 2 ]] || { echo "dev-deploy: --target requires a value" >&2; exit 1; }
             REMOTE="$2"; shift 2 ;;
         --units)
             [[ $# -ge 2 ]] || { echo "dev-deploy: --units requires a value" >&2; exit 1; }
@@ -113,7 +113,7 @@ echo "--- Syncing runtime/ to ${SSH_TARGET}:${RUNTIME_DST}"
 
 # shellcheck disable=SC2086
 rsync -avz --delete ${DRY} \
-    -e "ssh -o StrictHostKeyChecking=no" \
+    -e "ssh -o StrictHostKeyChecking=accept-new" \
     "${RUNTIME_SRC}" \
     "${SSH_TARGET}:${RUNTIME_DST}"
 
@@ -132,7 +132,7 @@ else
         printf 'Restarting %s...\n' "${unit}"
         # Treat a missing unit as a warning, not a fatal error — not all units
         # may be active on a partial dev image.
-        if ! ssh -o StrictHostKeyChecking=no "${SSH_TARGET}" \
+        if ! ssh -o StrictHostKeyChecking=accept-new "${SSH_TARGET}" \
                 "sudo systemctl restart ${unit}" 2>/dev/null; then
             printf '[WARN] %s: restart failed or unit not found — skipping\n' "${unit}" >&2
         else

--- a/scripts/flash-sd.sh
+++ b/scripts/flash-sd.sh
@@ -112,8 +112,8 @@ if [[ "${OS}" == "Linux" ]]; then
     # Extract the base device name (strip /dev/ prefix).
     dev_name="${DEV##*/}"
     # For mmcblk0p1, the sysfs removable entry lives under mmcblk0.
-    dev_base="${dev_name%%p[0-9]*}"
-    dev_base="${dev_base%%[0-9]}"
+    dev_base="${dev_name%%p[0-9]*}"   # strip mmcblk/nvme partition suffix (mmcblk0p1 -> mmcblk0)
+    dev_base="${dev_base%[0-9]*}"     # strip sata/usb partition suffix (sdb1 -> sdb); no-op for mmcblk0
     sysfs_removable="/sys/block/${dev_base}/removable"
     if [[ -f "${sysfs_removable}" ]]; then
         removable="$(cat "${sysfs_removable}")"

--- a/scripts/flash-sd.sh
+++ b/scripts/flash-sd.sh
@@ -113,7 +113,12 @@ if [[ "${OS}" == "Linux" ]]; then
     dev_name="${DEV##*/}"
     # For mmcblk0p1, the sysfs removable entry lives under mmcblk0.
     dev_base="${dev_name%%p[0-9]*}"   # strip mmcblk/nvme partition suffix (mmcblk0p1 -> mmcblk0)
-    dev_base="${dev_base%[0-9]*}"     # strip sata/usb partition suffix (sdb1 -> sdb); no-op for mmcblk0
+    # mmcblk/nvme names legitimately end in a digit (controller number); only strip
+    # trailing digits for sdX/loop-style names where the digit IS the partition number.
+    case "${dev_base}" in
+        mmcblk*|nvme*) : ;;
+        *) dev_base="${dev_base%%[0-9]*}" ;;
+    esac
     sysfs_removable="/sys/block/${dev_base}/removable"
     if [[ -f "${sysfs_removable}" ]]; then
         removable="$(cat "${sysfs_removable}")"

--- a/scripts/lib/boot-config.sh
+++ b/scripts/lib/boot-config.sh
@@ -59,7 +59,9 @@ write_boot_config() {
     local loop_dev mnt_boot mnt_a
     loop_dev="$(sudo losetup -f --show -P "${output_img}")"
 
-    # shellcheck disable=SC2329
+    # SC2329: function is invoked indirectly via trap (not a direct call site)
+    # SC2317: commands inside a trap-registered function appear unreachable to shellcheck
+    # shellcheck disable=SC2329,SC2317
     _bconf_cleanup() {
         sudo umount "${mnt_boot}" 2>/dev/null || true
         sudo umount "${mnt_a}" 2>/dev/null || true

